chatlearn/models/vllm_module_v2.py

```diff
@@ -0,0 +1,214 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""VLLM module"""
+
+import asyncio
+import inspect
+import os
+import sys
+from typing import Optional
+
+import torch
+from transformers import AutoTokenizer
+from vllm import SamplingParams
+from vllm.config import EngineConfig
+from vllm.executor.ray_utils import RayWorkerWrapper
+from vllm.engine.async_llm_engine import AsyncLLMEngine
+from vllm.engine.arg_utils import AsyncEngineArgs
+from vllm.usage.usage_lib import UsageContext
+from vllm.utils import FlexibleArgumentParser
+
+from chatlearn.utils.global_vars import set_vllm_actors
+from .torch_module import TorchModule
+
+
+class VLLMModuleV2(TorchModule, RayWorkerWrapper):
+    """VLLMModule is the class for vLLM models.
+
+    Args
+    ----
+    name : str
+        model name
+    """
+
+    def __init__(self, *args, **kwargs):
+        # avoid overwrite methods
+        methods_class1 = {method[0] for method in inspect.getmembers(TorchModule, predicate=inspect.isfunction)}
+        methods_class2 = {method[0] for method in inspect.getmembers(RayWorkerWrapper, predicate=inspect.isfunction)}
+        common_methods = methods_class1.intersection(methods_class2)
+        # common method is '__init__'
+        assert len(common_methods) == 1
+        TorchModule.__init__(self, *args)
+        self.local_rank = 0
+        os.environ['LOCAL_RANK'] = '0'
+        if 'worker_module_name' in kwargs and 'worker_class_name' in kwargs:
+            RayWorkerWrapper.__init__(self, **kwargs) # pylint: disable=non-parent-init-called
+        os.environ['VLLM_HOST_IP'] = self.get_address()
+        self.engine = None
+
+    def setup(self):
+        """Set up model and load checkpoint"""
+        tokenizer = AutoTokenizer.from_pretrained(self.model_args['tokenizer'])
+        tokenizer.tokenizer = tokenizer
+        self.tokenizer = tokenizer
+
+    def setup_vllm(self, workers):
+        # setup vllm engine in rank 0
+        os.environ['VLLM_HOST_IP'] = self.get_address()
+        set_vllm_actors(workers)
+        parser = FlexibleArgumentParser()
+        parser = AsyncEngineArgs.add_cli_args(parser)
+        backup_sys_argv = sys.argv
+        dtype = self.model_args.get("dtype", "bfloat16")
+        if self.model_args.get("fp16", False):
+            dtype = "float16"
+        vllm_sys_argv = ["",
+                         f"--model={self.model_args['load']}",
```

Comment:
must load model when new AsyncEngine? we needed lazy init model by parameter sync.

Reply:
lazy init model will be supported later in next pull request for training

Reply:
setup_vllm --> setup_vllm_eval ?

Reply:
setup_vllm is a common method for vllm, not only for eval only but also e2e